### PR TITLE
ARC-1291: Client Authentication for Token Endpoint

### PR DESCRIPTION
**Pull request by thomasrichner-oviva**

### CODE REVIEW

File: ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/ws/TokenEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      @FormParam("redirect_uri") String redirectUri,
+      @FormParam("client_id") String clientId,
+      @FormParam("client_assertion_type") String clientAssertionType,
+      @FormParam("client_assertion") String clientAssertion) {
+
+    if (!"authorization_code".equals(grantType)) {
+      return Response.status(Status.BAD_REQUEST).entity("bad 'grant_type': " + grantType).build();
+    }
+
+    authenticator.authenticate(new Request(clientId, clientAssertionType, clientAssertion));
```
Comment (thomasrichner-oviva) [inline ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/ws/TokenEndpoint.java:45]: This is the actual authentication.

File: ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/util/DiscoveryJwkSource.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.nio.charset.StandardCharsets;
+
+/** A JWK source that always fetches the latest JWKS from a given OpenID Discovery document URL */
+public class DiscoveryJwkSource<T extends SecurityContext> implements JWKSetSource<T> {
```
Comment (michelealbanese-oviva) [inline ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/util/DiscoveryJwkSource.java:25]:
```suggestion
public class DiscoveryJwkSetSource<T extends SecurityContext> implements JWKSetSource<T> {
```

more correct right?

File: ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/ws/OpenIdConfiguration.java [hunk 3]
```diff
@@ -71,4 +71,14 @@ public record OpenIdConfiguration(
...
-        List<String> idTokenEncryptionEncValuesSupported) {}
+        List<String> idTokenEncryptionEncValuesSupported,
+
+    // OPTIONAL. JSON array containing a list of Client Authentication methods supported by this
+    // Token Endpoint. The options are client_secret_post, client_secret_basic, client_secret_jwt,
+    // and private_key_jwt, as described in Section 9 of OpenID Connect Core 1.0 [OpenID.Core].
+    // Other authentication methods MAY be defined by extensions. If omitted, the default is
+    // client_secret_basic -- the HTTP Basic Authentication Scheme specified in Section 2.3.1 of
+    // OAuth 2.0 [RFC6749].
+    // See also: https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication
```
Comment (michelealbanese-oviva) [inline ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/ws/OpenIdConfiguration.java:82]: aren't you supporting `private_key_jwt` only though? I'd use that as a default instead

**Review by michelealbanese-oviva — APPROVED** [review]
looks good! some minor suggestions :)

File: ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/util/DiscoveryJwkSource.java (hunk 2 above)
Comment (thomasrichner-oviva) [reply ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/util/DiscoveryJwkSource.java:25]: Indeed! Leftover from a refactor :)

File: ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/ws/OpenIdConfiguration.java (hunk 3 above)
Comment (thomasrichner-oviva) [reply ehealthid-rp/src/main/java/com/oviva/ehealthid/relyingparty/ws/OpenIdConfiguration.java:82]: This is the generic model for an OpenIdConfiguration with the documentation from the spec. Yes, in reality for now it is always `private_key_jwt` but could easily be extended.
I also figured out the Keycloak actually just ignores it and need manual config :/

File: README.md [hunk 4]
```diff
@@ -100,29 +103,45 @@ Once the server is booted, it will:
 
 The discovery document can be used to configure the relying party in an existing identity provider.
 
+## Configure Identity Provider
+
+General:
```
[the remark this answers is not in the document]
Comment (thomasrichner-oviva) [reply README.md:108]: rephrasing

**Review by michelealbanese-oviva — APPROVED** [review]

**Pull request merged**